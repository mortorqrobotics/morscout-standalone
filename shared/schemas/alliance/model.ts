import { Schema } from "mongoose";
import IAlliance from "./interface";

export default new Schema<IAlliance>({
  teams: [
    {
      type: Schema.Types.ObjectId,
      ref: "Team",
      required: true
    }
  ]
});
