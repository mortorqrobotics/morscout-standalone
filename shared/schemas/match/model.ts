import { Schema } from "mongoose";
import IMatch from "./interface";

export default new Schema<IMatch>({
  redAlliance: {
    type: Schema.Types.ObjectId,
    ref: "Alliance",
    required: true
  },
  blueAlliance: {
    type: Schema.Types.ObjectId,
    ref: "Alliance",
    required: true
  },
  winner: {
    type: String,
    enum: ["Red", "Blue"]
  },
  startDate: {
    type: Date,
    required: true
  },
  endDate: {
    type: Date,
    required: true
  },
  Regional: {
    type: Schema.Types.ObjectId,
    ref: "Regional",
    required: true
  }
});
