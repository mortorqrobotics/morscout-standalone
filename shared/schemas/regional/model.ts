import { Schema } from "mongoose";
import { pointSchema, polygonSchema } from "../util/geojson";
import IRegional from "./interface";

const RegionalSchema: Schema<IRegional> = new Schema({
  name: {
    type: String,
    required: true
  },
  location: {
    type: pointSchema,
    required: true
  },
  district: polygonSchema,
  url: {
    type: String,
    validate: url => encodeURI(url) === url,
    required: true
  },
  eventType: {
    type: String,
    enum: [
      "regional",
      "district",
      "districtCompetition",
      "championshipDivision",
      "championshipFinals",
      "disitrictCompetitionDivision",
      "festivalOfChampions",
      "offseason",
      "preseason",
      "unlabeled"
    ],
    required: true
  },
  startDate: {
    type: Date,
    required: true
  },
  endDate: {
    type: Date,
    required: true
  },
  season: {
    type: Schema.Types.ObjectId,
    ref: "Season",
    required: true
  },
  key: {
    type: String,
    required: true,
    unique: true,
    index: true
  }
});

export default RegionalSchema;
