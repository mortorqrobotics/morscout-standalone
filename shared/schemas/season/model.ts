import { Schema } from "mongoose";
import ISeason from "./interface";

export default new Schema<ISeason>({
  year: {
    type: Number,
    validate: year => year >= 1989 && year <= new Date().getUTCFullYear() + 1,
    default: () => new Date().getUTCFullYear(),
    required: true
  },
  name: {
    type: String,
    required: true
  },

  buildStart: {
    type: Date,
    required: true
  },
  buildEnd: {
    type: Date,
    required: true
  },
  seasonStart: {
    type: Date,
    required: true
  },
  seasonEnd: {
    type: Date,
    required: true
  }
});
