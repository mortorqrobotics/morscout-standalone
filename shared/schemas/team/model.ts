import { Schema } from "mongoose";
import { pointSchema } from "../util/geojson";
import ITeam from "./interface";

export default new Schema<ITeam>({
  name: {
    // The team name
    type: String,
    required: true
  },
  number: {
    // Just the team number
    type: Number,
    validate: [
      num =>
        Number.isSafeInteger(num) &&
        Number.isFinite(num) &&
        !Number.isNaN(num) &&
        num > 0 &&
        num <= 9999,
      "{ Value } is not a legal team number"
    ],
    required: true
  },
  /**
   * Awards the team has earned
   * Format:
   * {
   *    year: ["Awards"]
   * }
   */
  awards: {
    type: Map,
    required: true,
    default: new Map()
  },
  // List of Registered Regionals
  registeredRegionals: [
    {
      type: Schema.Types.ObjectId,
      ref: "Regional"
    }
  ],
  location: {
    type: pointSchema,
    required: true
  },
  regional: {
    type: String
  }
});
