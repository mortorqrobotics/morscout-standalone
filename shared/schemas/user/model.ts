import * as mongoose from "mongoose";
import { compare, hash, genSalt } from "bcryptjs";
import IUser from "./interface";

let SALT_WORK_FACTOR = 10;

var userSchema = new mongoose.Schema<IUser>({
  username: {
    type: String,
    required: true,
    unique: true
  },
  password: {
    type: String,
    required: true,
    select: false
  },
  firstname: {
    type: String,
    required: true
  },
  lastname: {
    type: String,
    required: true
  },
  email: {
    type: String,
    required: true,
    unique: true
  },
  parentEmail: String,
  phone: {
    type: String, // changed from Number
    required: true,
    unique: true
  },
  profpicpath: String,
  team: {
    type: mongoose.Schema.Types.ObjectId,
    ref: "Team"
  },
  position: {
    type: String,
    enum: ["member", "leader", "mentor", "alumnus"]
  },
  scoutCaptain: {
    type: Boolean,
    default: false
  },
  mobileDeviceTokens: {
    type: [
      {
        type: String,
        required: true
      }
    ],
    default: []
  }
});

userSchema.pre<IUser>("save", function(next) {
  let capitalize = str => str[0].toUpperCase() + str.slice(1).toLowerCase();
  this.firstname = capitalize(this.firstname);
  this.lastname = capitalize(this.lastname);
  next();
});

userSchema.pre<IUser>("save", function(next) {
  let user = this;

  // if (!user.isModified("password")) return next(); Doesn't work

  genSalt(SALT_WORK_FACTOR, function(err: Error, salt: string) {
    hash(user.password, salt, function(err, hash) {
      user.password = hash;
      next();
    });
  });
});

userSchema.methods.comparePassword = function(candidatePassword) {
  let password = this.password;
  return new Promise((resolve, reject) =>
    compare(candidatePassword, password, function(err, isMatch) {
      if (err) {
        reject(err);
      } else {
        resolve(!!isMatch);
      }
    })
  );
};

export default userSchema;
