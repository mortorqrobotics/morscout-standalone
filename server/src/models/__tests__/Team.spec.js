import "mockingoose";
import model from "../Team";

test("Model is same throughout", async () => {
  expect.assertions(1);
  expect(
    model({
      model: (name, obj) => obj,
    }),
  ).toMatchSnapshot();
});
