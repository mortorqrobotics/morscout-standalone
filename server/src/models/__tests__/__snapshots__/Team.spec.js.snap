// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Model is same throughout 1`] = `
Schema {
  "$id": 1,
  "_indexes": Array [],
  "_userProvidedOptions": Object {},
  "aliases": Object {},
  "callQueue": Array [],
  "childSchemas": Array [],
  "inherits": Object {},
  "methodOptions": Object {},
  "methods": Object {},
  "nested": Object {
    "awards": true,
  },
  "obj": Object {
    "awards": Object {
      "Type": [Function],
    },
    "name": Object {
      "type": [Function],
    },
    "number": Object {
      "type": [Function],
      "validate": Object {
        "message": "{ Value } is not a legal team number",
        "validator": [Function],
      },
    },
  },
  "options": Object {
    "_id": true,
    "autoIndex": null,
    "bufferCommands": true,
    "capped": false,
    "discriminatorKey": "__t",
    "id": true,
    "minimize": true,
    "noId": false,
    "noVirtualId": false,
    "read": null,
    "shardKey": null,
    "strict": true,
    "typeKey": "type",
    "validateBeforeSave": true,
    "versionKey": "__v",
  },
  "paths": Object {
    "_id": ObjectId {
      "_index": null,
      "defaultValue": [Function],
      "getters": Array [],
      "instance": "ObjectID",
      "options": Object {
        "auto": true,
        "type": [Function],
      },
      "path": "_id",
      "setters": Array [
        [Function],
      ],
      "validators": Array [],
      Symbol(mongoose#schemaType): true,
    },
    "awards.Type": Map {
      "$__schemaType": Mixed {
        "_index": null,
        "getters": Array [],
        "instance": "Mixed",
        "options": Object {
          "type": Object {},
        },
        "path": "awards.Type.$*",
        "setters": Array [],
        "validators": Array [],
        Symbol(mongoose#schemaType): true,
        Symbol(mongoose:schema_mixed): true,
      },
      "$isSchemaMap": true,
      "_index": null,
      "getters": Array [],
      "instance": "Map",
      "options": Object {
        "type": [Function],
      },
      "path": "awards.Type",
      "setters": Array [],
      "validators": Array [],
      Symbol(mongoose#schemaType): true,
    },
    "awards.Type.$*": Mixed {
      "_index": null,
      "getters": Array [],
      "instance": "Mixed",
      "options": Object {
        "type": Object {},
      },
      "path": "awards.Type.$*",
      "setters": Array [],
      "validators": Array [],
      Symbol(mongoose#schemaType): true,
      Symbol(mongoose:schema_mixed): true,
    },
    "name": SchemaString {
      "_index": null,
      "enumValues": Array [],
      "getters": Array [],
      "instance": "String",
      "options": Object {
        "type": [Function],
      },
      "path": "name",
      "regExp": null,
      "setters": Array [],
      "validators": Array [],
      Symbol(mongoose#schemaType): true,
    },
    "number": SchemaNumber {
      "_index": null,
      "getters": Array [],
      "instance": "Number",
      "options": Object {
        "type": [Function],
        "validate": Object {
          "message": "{ Value } is not a legal team number",
          "validator": [Function],
        },
      },
      "path": "number",
      "setters": Array [],
      "validators": Array [
        Object {
          "message": "{ Value } is not a legal team number",
          "type": "user defined",
          "validator": [Function],
        },
      ],
      Symbol(mongoose#schemaType): true,
    },
  },
  "plugins": Array [],
  "query": Object {},
  "s": Object {
    "hooks": Kareem {
      "_posts": Map {},
      "_pres": Map {},
    },
  },
  "singleNestedPaths": Object {},
  "statics": Object {},
  "subpaths": Object {},
  "tree": Object {
    "_id": Object {
      "auto": true,
      "type": [Function],
    },
    "awards": Object {
      "Type": [Function],
    },
    "name": Object {
      "type": [Function],
    },
    "number": Object {
      "type": [Function],
      "validate": Object {
        "message": "{ Value } is not a legal team number",
        "validator": [Function],
      },
    },
  },
  "virtuals": Object {},
}
`;
